src/librbd/internal.cc

```diff
@@ -2797,6 +2822,58 @@ int validate_mirroring_enabled(ImageCtx *ictx) {
       }
     }
 
+    if (next_mirror_mode == cls::rbd::MIRROR_MODE_POOL) {
+      r = cls_client::mirror_mode_set(&io_ctx, cls::rbd::MIRROR_MODE_IMAGE);
+      if (r < 0) {
+        lderr(cct) << "Failed to set mirror mode to image: "
+          << cpp_strerror(r) << dendl;
+        return r;
+      }
+
+      map<string, string> images;
+      r = list_images_v2(io_ctx, &images);
+      if (r < 0) {
+        lderr(cct) << "Failed listing images: " << cpp_strerror(r) << dendl;
+        r = cls_client::mirror_mode_set(&io_ctx, current_mirror_mode);
+        if (r < 0) {
+          lderr(cct) << "Failed to rollback mirror mode: " << cpp_strerror(r)
+            << dendl;
+        }
+        return r;
+      }
+
+      for (const auto& img_pair : images) {
+        ImageCtx *img_ctx = new ImageCtx("", img_pair.second, nullptr, io_ctx,
```

Comment:
Might be nice to just directly query the image's features instead of fully opening the image (which requires a lot of back-and-forth). Use `librbd::cls_client::get_features` and pass it the header oid (computed from the image ID extracted from the directory via `librbd::util::header_name`) and CEPH_NOSNAP for the snap id.